includes/class-wc-payments-utils.php

```diff
@@ -583,4 +583,37 @@ public static function get_cached_minimum_amount( $currency ) {
 		$cached = get_transient( 'wcpay_minimum_amount_' . strtolower( $currency ) );
 		return (int) $cached ? (int) $cached : null;
 	}
+
+	/**
+	 * Locks an order for payment intent processing for 5 minutes.
+	 *
+	 * @param WC_Order $order  The order that is being paid.
+	 * @param string   $intent_id The id of the intent that is being processed.
+	 * @return bool            A flag that indicates whether the order is already locked.
+	 */
+	public static function lock_order_payment( $order, $intent_id = null ) {
+		$order_id       = $order->get_id();
+		$transient_name = 'wcpay_processing_intent_' . $order_id;
+		$processing     = get_transient( $transient_name );
+
+		// Block the process if the same intent is already being handled.
+		if ( '-1' === $processing || ( isset( $intent_id ) && $processing === $intent_id ) ) {
+			return true;
+		}
+
+		// Save the new intent as a transient, eventually overwriting another one.
+		set_transient( $transient_name, empty( $intent_id ) ? '-1' : $intent_id, 5 * MINUTE_IN_SECONDS );
+
+		return false;
```

Comment:
It seems a bit counterintuitive to me to return false after successfully locking, and true when locking fails. At least in the context of calling a method named `lock_order_payment`. 

It might be helpful if we either:

- rename this method to make it more clear that it is both checking and locking, or
- separate the checking functionality from the locking functionality

Reply:
Agree on return values @chickenn00dle 
I couldn't think of a situation where we'd want to check for a lock without immediately placing the lock, so was initially leaning against splitting `lock_order_payment`. But then again, coming up with a clearer function name was more difficult, so decided to go with splitting the function 😄 
This is adjusted in https://github.com/Automattic/woocommerce-payments/pull/3520/commits/d4ee61156db4327ce654916c7d7f75800984d07c